### ISSUE TITLE
Crash on v2.5.1-RC when scanning testdata vaccination and test certificate

### ISSUE THREAD

**Issue by Bubu** [corona-warn-app/cwa-app-android#3684]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [x] Bug is not already reported in another issue

## Technical details

- Device name: Pixel 4a
- Android version: Android 11
- App version: 2.5.1-RC (2943eff83e919812f9dffd0df2d6644898fa59e1)

## Describe the bug

When scanning the first 2 certificates from this testdata repo: https://github.com/eu-digital-green-certificates/dgc-testdata/tree/main/DE

Then navigating to the PersonDetailsFragment (might need navigating back to PersonOverview before, not sure) the app crashes.

## Steps to reproduce the issue

see above

## Expected behaviour

No crash.

## Additional context

Log:

```
07-11 11:21:19.830 31012 31012 E AndroidRuntime: FATAL EXCEPTION: main
07-11 11:21:19.830 31012 31012 E AndroidRuntime: Process: de.rki.coronawarnapp.test, PID: 31012
07-11 11:21:19.830 31012 31012 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean androidx.recyclerview.widget.RecyclerView$ViewHolder.shouldIgnore()' on a null object reference
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep3(RecyclerView.java:4161)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3862)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4404)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.google.android.material.appbar.HeaderScrollingViewBehavior.layoutChild(HeaderScrollingViewBehavior.java:148)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.google.android.material.appbar.ViewOffsetBehavior.onLayoutChild(ViewOffsetBehavior.java:43)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior.onLayoutChild(AppBarLayout.java:2003)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:918)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.coordinatorlayout.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1213)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:899)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:919)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.android.internal.policy.DecorView.onLayout(DecorView.java:784)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.View.layout(View.java:22844)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6389)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3475)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2943)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1948)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8177)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:972)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:796)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:731)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:957)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7664)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
07-11 11:21:19.830 31012 31012 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

EXPOSUREAPP-8336

**Comment by MikeMcC399** [comment]
@Bubu 
Nice catch! I can reproduce this on Android 8

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.rki.coronawarnapp.test, PID: 25075
    java.lang.IllegalStateException: Two different ViewHolders have the same stable ID. Stable IDs in your adapter MUST BE unique and SHOULD NOT change.
     ViewHolder 1:VaccinationCertificateCard{b67a9ab position=3 id=346916434, oldPos=-1, pLpos:-1} 
     View Holder 2:TestCertificateCard{1f044f2 position=2 id=346916434, oldPos=-1, pLpos:-1} androidx.recyclerview.widget.RecyclerView{a5ce9c7 VFED..... .F....ID 0,168-1080,1812 #7f0a059f app:id/recycler_view_certificates_list}, adapter:de.rki.coronawarnapp.covidcertificate.person.ui.details.PersonDetailsAdapter@73c04e, layout:androidx.recyclerview.widget.LinearLayoutManager@34d56f, context:de.rki.coronawarnapp.ui.main.MainActivity@25c099a
```

## Steps to reproduce

1. Tap "Certificates"
2. Tap "ADD CERTIFICATE"
3. From page https://github.com/eu-digital-green-certificates/dgc-testdata/tree/main/DE scan vaccination certificate (1)
4. Tap back arrow, tap X
5. Tap "ADD CERTIFICATE"
6. From page above scan test certificate (2)
7. Tap back arrow
8. Scroll down and tap Vaccination Certificate card at bottom
9. Tap back arrow
10. App crashes

**Comment by vaubaehn** [comment]
This smells very much tested to https://github.com/corona-warn-app/cwa-app-android/pull/3680

Any way to get this into 2.5.x, @harambasicluka and @d4rken ?

**Comment by vaubaehn** [comment]
@Bubu @MikeMcC399 
As far as I could see, both certificates share the same cert id. As this condition shouldn't occur often in practice, maybe fixing with 2.6.x like it is planned for now could be enough m

**Comment by Bubu** [comment]
@vaubaehn 

> As this condition shouldn't occur often in practice, maybe fixing with 2.6.x like it is planned for now could be enough?

I suppose that is fine if this case is not actually supposed to occur in the wild. 

There's some sort of (theoretical) DOS attack vector here but see conclusion below:

The certificates aren't cryptographically checked when scanning, so anyone can create a certificate with a clashing ID (if you know the id of the original cert). When you get someone to scan this fake cert it'll render that particular screen in the app useless.

But even though that's a bit annoying the system as whole would remain functional. (The main cert is shown in the previous screen and you can still add new certs). So that attack seems a) quite unlikely b) not very effective.

**Comment by d4rken** [comment]
@vaubaehn's :nose: is correct. This will be fixed by #3680: Two certificate having the same `certificateId` and the `RecyclerView` optimizations require `stableId = certificateId` to be unique.

> Any way to get this into 2.5.x, @harambasicluka and @d4rken ?

We had discussed this and decided against it as the certificate IDs "should" be unique, it does not warrant cutting another RC for 2.5.x and delay its release.

It's a bit unfortunate that the demo certificates re-use the same ID, especially for vaccination certificate chains, which @fynngodau [previously raised already](https://github.com/eu-digital-green-certificates/dgc-testdata/issues/326).

I think iOS currently prevents adding certificates with the same ID? Due to differences in storage, Android only checks the IDs per certificate-type.

We are evaluating to check the certificateId against all types on Android too, and preventing adding them if there is a collision. Though this is kinda a catch22: If the IDs are in practice not as unique as they should be, then "valid" certificates with duplicate IDs have been issued... when this happened with the date stuff, the checks were relaxed...

In theory they should be unique, in practice :shrug: :

> [It is the Member states’ responsibility to come up with the mechanism for generating and indexing the forementioned single unique vaccination identifiers.](https://ec.europa.eu/health/sites/default/files/ehealth/docs/vaccination-proof_interoperability-guidelines_en.pdf)

@Bubu DDOS case is an interesting point too. It should be fixed by #3680 too, other logic across certificates types doesn't rely on the `certificateId` being unique.

**Comment by MikeMcC399** [comment]
All 4 test certificates in https://github.com/eu-digital-green-certificates/dgc-testdata/tree/main/DE/2DCode/raw use 
                "ci": "URN:UVCI:01DE/IZ12345A/5CWLU12RNOB9RXSEOP6FG8#W"
🙁

There hasn't been any response to @fynngodau's issue raised one month ago in https://github.com/eu-digital-green-certificates/dgc-testdata/issues/326! Maybe @SZoerner (IBM) would be able to comment?

**Comment by MikeMcC399** [comment]
I cannot reproduce the crash using the steps from https://github.com/corona-warn-app/cwa-app-android/issues/3684#issuecomment-877782403 in the released versions CWA Android 2.5.1 or 2.6.1.

In CWA Android 2.7.1 the certificates could not be successfully scanned. Error `HC_DSC_NO_MATCH` was shown, since the certificates are not signed, so also no crash in that configuration.

I re-built 2.5.1-RC1 deviceForTestersDebug (https://github.com/corona-warn-app/cwa-app-android/commit/2943eff83e919812f9dffd0df2d6644898fa59e1) and I could not produce the crash there either, even though it is the same device where I got a crash 6 weeks ago! I can't explain that.

However, the requirement from CWA Android 2.7.1 that certificates must be correctly signed, and that a correctly signed certificate would always have a unique ci means that this issue can no longer occur. Unless somebody knows of a good reason, this issue can probably therefore be closed now.

**Comment by MikeMcC399** [comment]
@dsarkar 
I suggest to close this issue.
